clubs/mixins.py

```diff
@@ -0,0 +1,241 @@
+
+        # turn all entries into dict configs
+        for i, m2m in enumerate(m2m_to_save):
+            if not isinstance(m2m, dict):
+                m2m_to_save[i] = {
+                    'field': m2m,
+                    'create': False
+                }
+
+        # remove m2m from validated data and save
+        m2m_lists = {}
+        for m2m in m2m_to_save:
+            create = m2m.get('create', False)
+            field_name = m2m['field']
+
+            field = self.fields[field_name]
+            if isinstance(field, serializers.ListSerializer):
+                m2m['many'] = True
+                model = field.child.Meta.model
+                m2m_lists[field_name] = []
+                items = self.validated_data.pop(field_name, None)
+                if items is None:
+                    continue
+                for item in items:
+                    m2m_lists[field_name].append(self._lookup_item(model, field_name, item, create))
+            else:
+                m2m['many'] = False
+                model = field.Meta.model
+                item = self.validated_data.pop(field_name, None)
+                m2m_lists[field_name] = self._lookup_item(model, field_name, item, create)
+
+        obj = super(ManyToManySaveMixin, self).save()
+
+        # link models to this model
+        updates = []
+        for m2m in m2m_to_save:
+            field = m2m['field']
+            value = m2m_lists[field]
+            if m2m['many']:
+                getattr(obj, field).set(value)
+            else:
+                setattr(obj, field, value)
+                updates.append(field)
+
+        if updates:
+            obj.save(update_fields=updates)
+
+        return obj
+
+
+class XLSXFormatterMixin(object):
+    """
+    Mixin for views that formats xlsx output to a more readable format.
+    Will only apply to views that implement a get_serializer_class method.
+
+    You can insert "format_{field}_for_spreadsheet" methods in your serializer class
+    that accept a single argument (the cell value) and returns the formatted value.
+
+    Changes the default filename to include the date and time of creation.
+    Changes the default column header to be bolded.
+    """
+
+    def _format_header_value(self, key):
+        """
+        Format the text displayed in the column header.
+        """
+        return key.replace('_', ' ').title()
+
+    def _many_to_many_individual_formatter(self, value):
+        """
+        Take in an individual object generated by a ManyToMany serializer and format it.
+        If there is a single field in the dictionary, return that. Otherwise, return the
+        first field that is not the id field.
+        """
+        if isinstance(value, dict):
+            if len(value) == 1:
+                return next(value.values())
+            elif len(value) > 1:
+                return next(v for k, v in value.items() if not k == 'id')
```

Comment:
is this deterministic? I know in C++ iterating through maps causes different orderings pretty frequently, not sure how Python handles it.

Reply:
This should actually be deterministic, because we're iterating over an `OrderedDict` instead of a normal dictionary. Otherwise, you would be right, this would not be deterministic!